backend/regalloc/regalloc_invariants.ml

```diff
@@ -0,0 +1,218 @@
+    Array.fold_left regs ~init:0 ~f:(fun acc reg ->
+        match reg.Reg.loc with
+        | Unknown | Reg _ | Stack (Incoming _ | Outgoing _ | Domainstate _) ->
+          acc
+        | Stack (Local _) -> succ acc)
+  in
+  let arch_constraints (id : Instruction.id) (desc : Cfg.basic)
+      (arg : Reg.t array) (res : Reg.t array) : unit =
+    match Config.architecture with
+    (* CR xclerc for xclerc: what about cross-compilation? *)
+    | "amd64" | "arm64" -> (
+      let num_locals = num_stack_locals arg + num_stack_locals res in
+      match desc with
+      | Op (Spill | Reload) ->
+        (* CR xclerc for xclerc: should check arg/res according to spill/reload,
+           rather than the total number. *)
+        if num_locals > 1
+        then
+          fatal "instruction %d is a move and refers to %d spilling slots" id
+            num_locals
+      | _ -> ())
+    | arch -> fatal "unsupported architecture %S" arch
+  in
+  let register_classes_must_be_consistent (id : Instruction.id) (reg : Reg.t) :
+      unit =
+    match reg.Reg.loc with
+    | Reg phys_reg ->
+      let phys_reg = Proc.phys_reg phys_reg in
+      if not (same_reg_class reg phys_reg)
+      then
+        fatal
+          "instruction %d assigned %a to %a but they are in different classes"
+          id Printmach.reg reg Printmach.reg phys_reg
+    | Stack _ | Unknown -> ()
+  in
+  let register_classes_must_be_consistent (id : Instruction.id)
+      (regs : Reg.t array) : unit =
+    ArrayLabels.iter regs ~f:(register_classes_must_be_consistent id)
+  in
+  Cfg_with_layout.iter_instructions cfg_with_layout
+    ~instruction:(fun instr ->
+      let id = instr.id in
+      registers_must_not_be_unknown id instr.arg;
+      registers_must_not_be_unknown id instr.res;
+      arch_constraints id instr.desc instr.arg instr.res;
+      register_classes_must_be_consistent id instr.arg;
+      register_classes_must_be_consistent id instr.res)
+    ~terminator:(fun term ->
+      let id = term.id in
+      registers_must_not_be_unknown id term.arg;
+      registers_must_not_be_unknown id term.res;
+      register_classes_must_be_consistent id term.arg;
+      register_classes_must_be_consistent id term.res);
+  let fun_num_stack_slots =
+    (Cfg_with_layout.cfg cfg_with_layout).fun_num_stack_slots
+  in
+  let reg_class = ref 0 in
+  Array.iter2 max_stack_slots fun_num_stack_slots ~f:(fun max_slot num_slots ->
+      (* CR-soon xclerc for xclerc: make the condition stricter. The present
+         condition ensure safety: no access out of bounds (i.e. to another
+         frame), but could be made stricter to ensure optimiality (i.e. we use
+         every element from the frame). *)
+      if max_slot >= num_slots
+      then
+        fatal
+          "register class %d has a max slot of %d, but the number of slots is \
+           %d"
+          !reg_class max_slot num_slots;
+      incr reg_class)
+
+let postcondition_liveness : Cfg_with_liveness.t -> unit =
```

Comment:
Sorry, I meant to comment on that in the pull request description. You are right, and the reason why there is a new function rather than augmenting the existing one is that it was making the interface of the validator a bit awkward.